inductor: align cpp floordiv with python floordiv for dyanmic shape path

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102068

This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @ezyang @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

XiaobingSuper committed:
inductor: align cpp floordiv with python floordiv for dyanmic shape path

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc ezyang voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc ezyang voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -218,44 +218,58 @@ def _print_ModularIndexing(self, expr):
         assert len(expr.args) == 1
-        return f"std::floor({self._print(expr.args[0])})"
+        r = f"std::floor({self._print(expr.args[0])})"
+        return f"static_cast<{INDEX_TYPE}>({r})" if expr.is_integer else r
```
Please replace `INDEX_TYPE` with another name because it does not represent an index value.

XiaobingSuper replied:
Change to ```long```.

EikanWang commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -218,44 +218,58 @@ def _print_ModularIndexing(self, expr):
-        return f"std::floor({x} / {div})"
+        if expr.is_integer:
+            return f"div_floor_internal({x}, {div})"
+        return (
```
`(` is useless.

XiaobingSuper replied:
This ```lintrunner``` add that.

EikanWang reviewed (changes requested):
Please add unit test cases.

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc ezyang voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc ezyang voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -218,44 +218,58 @@ def _print_ModularIndexing(self, expr):
+            r = f"std::sqrt({base})"
+        else:
+            assert exp.is_integer
         exp = int(exp)
```
updated.

XiaobingSuper commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -218,44 +218,58 @@ def _print_ModularIndexing(self, expr):
         x = self.paren(self.doprint(x))
         div = self.paren(self.doprint(div))
-        return f"std::floor({x} / {div})"
+        if expr.is_integer:
```
We need it, see the https://godbolt.org/z/z39onxdov, the original implementation gets a wrong value.

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc ezyang voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc jgong5 mingfeima sanchitintel ashokei jingxu10 ezyang voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc jgong5 mingfeima sanchitintel ashokei jingxu10 ezyang voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc jgong5 mingfeima sanchitintel ashokei jingxu10 ezyang voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc jgong5 mingfeima sanchitintel ashokei jingxu10 ezyang voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -106,6 +107,46 @@ inline float flag_to_float_scalar(T src) {
+}
+
+template <>
+inline long div_floor_internal(long a, long b) {
```
Yes, updated.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -216,53 +216,70 @@ def _print_Integer(self, expr):
 
 def cexpr_index(index):
-    return f"static_cast<{INDEX_TYPE}>({cexpr(index)})"
+    return f"static_cast<long>({cexpr(index)})"
```
We use `INDEX_TYPE` for the cpp type of indices. I would suggest we keep this, not hard-coding `long` here. Also, if we are sure the functions like `ModularIndexing` always returns an index, I would suggest we use `INDEX_TYPE` instead of `long` too.

XiaobingSuper replied:
Ok, updated.

XiaobingSuper commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -218,44 +218,58 @@ def _print_ModularIndexing(self, expr):
         mod = self.paren(self.doprint(mod))
-        if div != "1":
-            x = f"({x} / {div})"
+        if div != "1" and div != "1L" and div != "1.0":
```
Yes, changed.

XiaobingSuper commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -106,6 +107,46 @@ inline float flag_to_float_scalar(T src) {
+
+// Copy from aten/src/ATen/native/cpu/BinaryOpsKernel.cpp.
+template <typename scalar_t>
+inline scalar_t div_floor_internal(scalar_t a, scalar_t b) __ubsan_ignore_float_divide_by_zero__ {
```
Yes, updated.

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc jgong5 mingfeima sanchitintel ashokei jingxu10 ezyang voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc jgong5 mingfeima sanchitintel ashokei jingxu10 ezyang voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc jgong5 mingfeima sanchitintel ashokei jingxu10 ezyang voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc jgong5 mingfeima sanchitintel ashokei jingxu10 ezyang voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper commented:
@pytorchbot merge -r

XiaobingSuper committed:
Update on "inductor: align cpp floordiv with python floordiv for dyanmic shape path"


This PR does the following things:

- Align the C++ behavior with Python for FloorDiv.
- Always return expr dtype for some ops which not use expr's dtype to do the computation.

After this PR, TIMM ```levit_128``` and ```volo_d1_224``` accuracy tests can be passed for dynamic shape path.


cc jgong5 mingfeima sanchitintel ashokei jingxu10 ezyang voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]
